core/src/main/scala/kafka/coordinator/group/GroupCoordinator.scala

```diff
@@ -613,16 +624,29 @@ class GroupCoordinator(val brokerId: Int,
         // The group is only using Kafka to store offsets.
         // Also, for transactional offset commits we don't need to validate group membership and the generation.
         groupManager.storeOffsets(group, memberId, offsetMetadata, responseCallback, producerId, producerEpoch)
-      } else if (group.is(CompletingRebalance)) {
-        responseCallback(offsetMetadata.mapValues(_ => Errors.REBALANCE_IN_PROGRESS))
       } else if (!group.has(memberId)) {
         responseCallback(offsetMetadata.mapValues(_ => Errors.UNKNOWN_MEMBER_ID))
       } else if (generationId != group.generationId) {
         responseCallback(offsetMetadata.mapValues(_ => Errors.ILLEGAL_GENERATION))
       } else {
-        val member = group.get(memberId)
-        completeAndScheduleNextHeartbeatExpiration(group, member)
-        groupManager.storeOffsets(group, memberId, offsetMetadata, responseCallback)
+        group.currentState match {
+          case Stable | PreparingRebalance =>
+            // During PreparingRebalance phase, we still allow a commit request since we rely
+            // on heartbeat response to eventually notify the rebalance in progress signal to the consumer
+            val member = group.get(memberId)
+            completeAndScheduleNextHeartbeatExpiration(group, member)
+            groupManager.storeOffsets(group, memberId, offsetMetadata, responseCallback)
+
+          case CompletingRebalance =>
```

Comment:
I think you mentioned there are some implication for KIP-429. In theory, if the consumer knows it will continue to own some partitions, it _could_ continue to commit offsets before it has the assignment for the latest generation. It seems we should either explicitly allow this or treat it as an illegal state?

Reply:
I've thought about that, and it is actually a bit complicated for consumer to do so. Because if you get a retriable error code from commit response in commitSync (in commitAsync we do not retry but just set the error back to user), the re-try logic is:

1) ensureCoordinatorReady.
2) resend commit request

Which means that we would not re-join group since we do not check `ensureActiveGroup` at all. On the other hand, `ensureActiveGroup` makes sure that the whole rebalance is done including `onPartitionRevoked / onPartitionAssigned` callback, and therefore some partitions may already be revoked -- and in practice we are going to commit offsets in `onPartitionRevoked` successfully since the group state would be in `stable` anyways. Hence for this `commitSync`, it is better still fail and throw.

However, I'm thinking that instead of throwing CommitFailed, we can instead throw a different exception (e.g. we can just throw `RebalanceInProgressException` all the way to the caller). In callers like Streams then, we should handle CommitFailed and RebalanceInProgress differently: e.g. for the former we treat it as "task being migrated", while for the latter, since we know we are going to call `commitSync` again in `onPartitionsRevoked` we can just swallow it and proceed.

Reply:
Note though, it would require a minor public API change that `commitSync` could throw `RebalanceInProgressException`. I'd just piggy back with KIP-429.

Reply:
Ok, let's keep it as is.